docs/cyberprobe-cli.1.in
.TH CYBERPROBE-CLI "1" "August 2016" "@PACKAGE@" "@VERSION@" "User Commands"
.SH NAME
cyberprobe-cli \- administration utility for cyberprobe management service.
.SH SYNOPSIS
.B cyberprobe-cli
\fIHOST\fR
\fIPORT\fR
.SH DESCRIPTION
.B cyberprobe-cli
administrates a running
.B cyberprobe
via the management service.

The first argument specifies a hostname, the second the port number to
connect to.  Upon connection, there is an immediate prompt for username
and password, and once authenticated, a command-line dialogue takes place.
.SS "Exit status:"
.TP
0
if OK,
.TP
non-zero
on error
.SH "EXAMPLE USAGE"
cyberprobe-cli localhost 8888
.SH COMMANDS
The following commands are supported by 'cyberprobe-cli':
.TP
.B add endpoint HOST PORT TYPE
Adds a delivery endpoint.  HOST specifies the delivery host.
PORT specifies TCP port to deliver to. TYPE
can be one of
.B nhis
or
.B etsi
for delivery protocol.
.TP
.B add interface INTERFACE DELAY [FILTER]
Adds an interface for packet sniffing.
INTERFACE specifies the network interface name. DELAY specifies
delay between packet acquisiton and delivery.  Defaults to
zero.  FILTER is optional, species a filter to be applied for positive
selection of packets, in BPF / libpcap format.
.TP
.B add parameter KEY VALUE
Adds a parameter.  KEY is parameter key, VALUE is parameter value.
.TP
.B add target LIID PROTOCOL ADDRESS
Adds an address target for packet capture.
LIID specifies the LIID / device identifier.
PROTOCOL specifies address protocol, one of
.B ip4
or
.B ip6.
ADDRESS specifies address value, in IPv4 or IPv6 format, according to the
PROTOCOL value.
.TP
.B help
Displays help.
.TP
.B quit
Causes the client to close the connection and terminate.
.TP
.B remove endpoint HOST PORT TYPE
Removes an endpoint added through the
.B add endpoint
command.  The
HOST, PORT and TYPE values are the same as for
.B add endpoint.
.TP
.B remove interface INTERFACE DELAY [FILTER]
Removes an interface added through the
.B add interface
command.
The INTERFACE, DELAY and FILTER values are the same as for
.B add interface.
.TP
.B remove paramter KEY VALUE
Removes a paramter added through the
.B add parameter
command.  The
KEY and VALUE values are the same as for
.B remove parameter.
.TP
.B remove target PROTOCOL ADDRESS
Removes a target added through the
.B add target
command.  The
PROTOCOL and ADDRESS values are the same as for
.B add target.
.TP
.B show endpoints
Displays a table showing endpoints.
.TP
.B show interfaces
Displays a table showing interfaces.
.TP
.B show parameters
Displays a table showing parameters.
.TP
.B show targets
Displays a table showing targets.
.SH AUTHOR
Part of the Cyberprobe project.
.SH "REPORTING BUGS"
Visit <https://github.com/cybermaggedon/cyberprobe> for support.
.SH COPYRIGHT
Copyright \(co 2013-2017 Cyber MacGeddon
.br
Copyright \(co 2013-2017 Trust Networks Inc.
.br
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR cyberprobe (1)
.PP
Further documentation and a quickstart guide is available in
/usr/share/doc/cyberprobe.  The command
.IP
.B info cyberprobe
.PP
accesses the complete manual.


ChangeLog
2015-02-02  Cyber MacGeddon  <cybermaggedon@gmail.com>

	* Updated to version 0.61.

	* Fixed HTTP crashing problem in cybermon.

	* Fixed documentation paths.

	* Changed from use of cybox hostname to domainname for STIX handling.
	  This seems to be the more 'correct' approach.

2013-10-01  Cyber MacGeddon  <cybermaggedon@gmail.com>

	* Added autoconf support.

	* Added variant LUA compilation support for 5.1 and 5.2.

	* Now compiles on a MacBook!



src/cyberprobe-cli.C

#include "readline.h"
#include <iostream>
#include <sstream>
#include <cybermon/socket.h>
#include <vector>
#include <boost/regex.hpp>
#include <iomanip>

std::vector<std::string> commands;
std::vector<std::string> add_commands;
std::vector<std::string> remove_commands;
std::vector<std::string> show_commands;
std::vector<std::string> classes;

std::vector<std::string>::iterator table_pos;
std::vector<std::string>::iterator table_end;

std::string get_status(const std::string& line)
{
    int pos = line.find(" ");
    if (pos != -1)
	return line.substr(0, pos);
    else
	return "";
}

bool cmd_data(tcpip::tcp_socket& sock, const std::string& cmd, 
	      std::string& result) 
{

    sock.write(cmd + "\n");

    std::string line;
    sock.readline(line);
    if (get_status(line) != "201") {
	std::cout << line << std::endl;
	return false;
    }
    
    sock.readline(line);
    std::istringstream buf(line);
    int len;
    buf >> std::dec >> len;

    sock.read(result, len);

    return true;

}
    
void cmd_do(tcpip::tcp_socket& s, const std::string& cmd)
{

    s.write(cmd + "\n");

    std::string line;
    s.readline(line);

    if (get_status(line) != "200")
	std::cerr << line << std::endl;

}

void cmd_do2(tcpip::tcp_socket& s, const std::string& c1, 
	     const std::string& c2)
{
    std::string cmd = c1 + " " + c2;
    cmd_do(s, cmd);
}

void cmd_do3(tcpip::tcp_socket& s, const std::string& c1, 
	     const std::string& c2, const std::string& c3)
{
    std::string cmd = c1 + " " + c2 + " " + c3;
    cmd_do(s, cmd);
}

void cmd_do4(tcpip::tcp_socket& s, const std::string& c1, 
	     const std::string& c2, const std::string& c3, 
	     const std::string& c4)
{
    std::string cmd = c1 + " " + c2 + " " + c3 + " " + c4;
    cmd_do(s, cmd);
}

void cmd_do5(tcpip::tcp_socket& s, const std::string& c1, 
	     const std::string& c2, const std::string& c3, 
	     const std::string& c4, const std::string& c5)
{
    std::string cmd = c1 + " " + c2 + " " + c3 + " " + c4 + " " + c5;
    cmd_do(s, cmd);
}

void cmd_help()
{
    std::cerr << "Not implemented. :) " << std::endl;
}

void cmd_endpoints(tcpip::tcp_socket& sock) 
{

    std::string result;

    cmd_data(sock, "endpoints", result);

    std::cout.setf(std::ios::left);

    std::cout << std::setw(40) << "Hostname"
	      << std::setw(8) << "Port"
	      << std::setw(10) << "Type"
	      << std::endl;
    
    std::cout << std::setw(40) << "--------"
	      << std::setw(8) << "----"
	      << std::setw(10) << "----"
	      << std::endl;
    
    while (1) {

	int pos = result.find(":");
	if (pos == -1) break;
	std::string hostname = result.substr(0, pos);
	result = result.substr(pos + 1);

	pos = result.find(":");
	if (pos == -1) break;
	std::string port = result.substr(0, pos);
	result = result.substr(pos + 1);

	pos = result.find(":");
	if (pos == -1) break;
	std::string type = result.substr(0, pos);
	result = result.substr(pos + 1);

	pos = result.find("\n");
	if (pos == -1) break;
	std::string desc = result.substr(0, pos);
	result = result.substr(pos + 1);

	std::cout << std::setw(40) << hostname
		  << std::setw(8) << port
		  << std::setw(10) << type
		  << std::endl;

    }
    
}

void cmd_targets(tcpip::tcp_socket& sock) 
{

    std::string result;
    cmd_data(sock, "targets", result);

    std::cout.setf(std::ios::left);

    std::cout << std::setw(20) << "LIID"
	      << std::setw(8) << "Class"
	      << std::setw(30) << "Address"
	      << std::endl;
    
    std::cout << std::setw(20) << "----"
	      << std::setw(8) << "-----"
	      << std::setw(30) << "-------"
	      << std::endl;
    
    while (1) {

	int pos = result.find(":");
	if (pos == -1) break;
	std::string liid = result.substr(0, pos);
	result = result.substr(pos + 1);

	pos = result.find(":");
	if (pos == -1) break;
	std::string cls = result.substr(0, pos);
	result = result.substr(pos + 1);

	pos = result.find("\n");
	if (pos == -1) break;
	std::string addr = result.substr(0, pos);
	result = result.substr(pos + 1);

	std::cout << std::setw(20) << liid
		  << std::setw(8) << cls
		  << std::setw(30) << addr
		  << std::endl;

    }
    
}

void cmd_interfaces(tcpip::tcp_socket& sock) 
{

    std::string result;
    cmd_data(sock, "interfaces", result);

    std::cout.setf(std::ios::left);

    std::cout << std::setw(20) << "Interface"
	      << std::setw(8) << "Delay"
	      << std::setw(50) << "Filter"
	      << std::endl;
    
    std::cout << std::setw(20) << "---------"
	      << std::setw(8) << "-----"
	      << std::setw(50) << "------"
	      << std::endl;
    
    while (1) {

	int pos = result.find(":");
	if (pos == -1) break;
	std::string interface = result.substr(0, pos);
	result = result.substr(pos + 1);

	pos = result.find(":");
	if (pos == -1) break;
	std::string delay = result.substr(0, pos);
	result = result.substr(pos + 1);

	pos = result.find("\n");
	if (pos == -1) break;
	std::string filter = result.substr(0, pos);
	result = result.substr(pos + 1);

	std::cout << std::setw(20) << interface
		  << std::setw(8) << delay
		  << std::setw(50) << filter
		  << std::endl;

    }
    
}

void cmd_parameters(tcpip::tcp_socket& sock) 
{

    std::string result;
    cmd_data(sock, "parameters", result);

    std::cout.setf(std::ios::left);

    std::cout << std::setw(30) << "Key"
	      << std::setw(45) << "Value"
	      << std::endl;
    
    std::cout << std::setw(30) << "---"
	      << std::setw(45) << "-----"
	      << std::endl;
    
    while (1) {

	int pos = result.find(":");
	if (pos == -1) break;
	std::string key = result.substr(0, pos);
	result = result.substr(pos + 1);

	pos = result.find("\n");
	if (pos == -1) break;
	std::string value = result.substr(0, pos);
	result = result.substr(pos + 1);

	std::cout << std::setw(30) << key
		  << std::setw(45) << value
		  << std::endl;

    }
    
}

bool generator(const std::string& partial, std::string& match) 
{
    
    while (table_pos != table_end) {

	if (table_pos->substr(0, partial.size()) == partial) {
	    match = *table_pos;
	    table_pos++;
	    return true;
	}
	
	table_pos++;
	
    }

    readline::completion_over();

    return false;

}

void completer(const std::vector<std::string>& tokens,
	       const std::string& token,
	       int cur_token,
	       std::vector<std::string>& completions)
{

    if (cur_token == 0) {
	table_pos = commands.begin();
	table_end = commands.end();
	readline::make_completions(completions, token, generator);
	return;
    }

    if ((cur_token == 1) && (tokens[0] == "add")) {
    	table_pos = add_commands.begin();
    	table_end = add_commands.end();
	readline::make_completions(completions, token, generator);
    }

    if ((cur_token == 1) && (tokens[0] == "remove")) {
    	table_pos = remove_commands.begin();
    	table_end = remove_commands.end();
	readline::make_completions(completions, token, generator);
    }

    if ((cur_token == 1) && (tokens[0] == "show")) {
    	table_pos = show_commands.begin();
    	table_end = show_commands.end();
	readline::make_completions(completions, token, generator);
    }

    if ((cur_token > 1) && 
	(tokens[0] == "add") && (tokens[1] == "interface")) {
	std::cerr << std::endl << "  Usage:" << std::endl;
	std::cerr << "    add interface <interface> <delay> <filter>" 
		  << std::endl;
	readline::completion_over();
	readline::force_display_update();
	return;
    }

    if ((cur_token > 1) && 
	(tokens[0] == "remove") && (tokens[1] == "interface")) {
	std::cerr << std::endl << "  Usage:" << std::endl;
	std::cerr << "    remove interface <interface> <delay> <filter>" 
		  << std::endl;
	readline::completion_over();
	readline::force_display_update();
	return;
    }

    if ((cur_token > 1) && 
	(tokens[0] == "add") && (tokens[1] == "endpoint")) {
	std::cerr << std::endl << "  Usage:" << std::endl;
	std::cerr << "    add endpoint <host> <port> <type>" 
		  << std::endl;
	readline::completion_over();
	readline::force_display_update();
	return;
    }

    if ((cur_token > 1) && 
	(tokens[0] == "remove") && (tokens[1] == "endpoint")) {
	std::cerr << std::endl << "  Usage:" << std::endl;
	std::cerr << "    remove endpoint <host> <port> <type>" 
		  << std::endl;
	readline::completion_over();
	readline::force_display_update();
	return;
    }

    if ((cur_token == 3) &&
	(tokens[0] == "add") && (tokens[1] == "target")) {
    	table_pos = classes.begin();
    	table_end = classes.end();
	readline::make_completions(completions, token, generator);
	return;
    }

    if ((cur_token > 1) &&
	(tokens[0] == "add") && (tokens[1] == "target")) {
	std::cerr << std::endl << "  Usage:" << std::endl;
	std::cerr << "    add target <liid> [ipv4|ipv6] <address>" 
		  << std::endl;
	readline::completion_over();
	readline::force_display_update();
	return;
    }

    if ((cur_token == 2) &&
	(tokens[0] == "remove") && (tokens[1] == "target")) {
    	table_pos = classes.begin();
    	table_end = classes.end();
	readline::make_completions(completions, token, generator);
	return;
    }

    if ((cur_token > 1) &&
	(tokens[0] == "remove") && (tokens[1] == "target")) {
	std::cerr << std::endl << "  Usage:" << std::endl;
	std::cerr << "    remove target [ipv4|ipv6] <address>" 
		  << std::endl;
	readline::completion_over();
	readline::force_display_update();
	return;
    }

    if ((cur_token > 1) &&
	(tokens[0] == "add") && (tokens[1] == "parameter")) {
	std::cerr << std::endl << "  Usage:" << std::endl;
	std::cerr << "    add parameter <key> <value>" 
		  << std::endl;
	readline::completion_over();
	readline::force_display_update();
	return;
    }

    if ((cur_token > 1) &&
	(tokens[0] == "remove") && (tokens[1] == "parameter")) {
	std::cerr << std::endl << "  Usage:" << std::endl;
	std::cerr << "    remove parameter <key> <value>" 
		  << std::endl;
	readline::completion_over();
	readline::force_display_update();
	return;
    }

    readline::completion_over();

}

int client(int argc, char** argv)
{

    if (argc != 3) {
	std::cerr << "Usage:" << std::endl
		  << "\tcyberprobe_cli host port" << std::endl;
	exit(1);
    }

    commands.push_back("add");
    commands.push_back("remove");
    commands.push_back("show");
    commands.push_back("quit");
    commands.push_back("help");

    add_commands.push_back("interface");
    add_commands.push_back("target");
    add_commands.push_back("endpoint");
    add_commands.push_back("parameter");

    remove_commands = add_commands;

    show_commands.push_back("interfaces");
    show_commands.push_back("targets");
    show_commands.push_back("endpoints");
    show_commands.push_back("parameters");

    classes.push_back("ipv4");
    classes.push_back("ipv6");

    setupterm(getenv("TERM"), 1, 0);

    std::string host = argv[1];
    
    std::istringstream buf(argv[2]);
    unsigned int port;
    buf >> port;

    tcpip::tcp_socket sock;
    sock.connect(host, port);

    std::cout << "Connected.  You must authenticate." << std::endl;

    while (1) {

	std::string user, password;

	readline::get_line("User: ", user);

	readline::get_password("Password: ", password);

	std::string cmd = "auth " + user + " " + password + "\n";

	sock.write(cmd);

	std::string response;
	sock.readline(response);

	std::string status;
	int pos = response.find(" ");
	if (pos != -1)
	    status = response.substr(0, pos);

	if (status == "200") {
	    std::cout << "Authentication successful." << std::endl;
	    break;
	} else
	    std::cout << response << std::endl;

    }

    while (1) {

	std::string s;
	readline::get_line_completing("> ", s, completer);

	static const boost::regex 
	    help(" *help *$", boost::regex::extended);

	boost::match_results<std::string::const_iterator> what;

	if (regex_search(s, help, boost::match_continuous)) {
	    cmd_help();
	    continue;
	}

	static const boost::regex 
	    endpoints(" *show +endpoints *$", boost::regex::extended);

	if (regex_search(s, endpoints, boost::match_continuous)) {
	    cmd_endpoints(sock);
	    continue;
	}

	static const boost::regex 
	    targets(" *show +targets *$", boost::regex::extended);

	if (regex_search(s, targets, boost::match_continuous)) {
	    cmd_targets(sock);
	    continue;
	}

	static const boost::regex 
	    interfaces(" *show +interfaces *$", boost::regex::extended);

	if (regex_search(s, interfaces, boost::match_continuous)) {
	    cmd_interfaces(sock);
	    continue;
	}

	static const boost::regex 
	    parameters(" *show +parameters *$", boost::regex::extended);

	if (regex_search(s, parameters, boost::match_continuous)) {
	    cmd_parameters(sock);
	    continue;
	}

	static const boost::regex 
	    add_interface(" *add +interface +([^ ]+) +([^ ]+) *$", 
			  boost::regex::extended);

	if (regex_search(s, what, add_interface, boost::match_continuous)) {
	    cmd_do3(sock, "add_interface", what[1], what[2]);
	    continue;
	}

	static const boost::regex 
	    remove_interface(" *remove +interface +([^ ]+) +([^ ]+) *$", 
			  boost::regex::extended);

	if (regex_search(s, what, remove_interface, boost::match_continuous)) {
	    cmd_do3(sock, "remove_interface", what[1], what[2]);
	    continue;
	}

	static const boost::regex 
	    add_target(" *add +target +([^ ]+) +([^ ]+) +([^ ]+) *$", 
			  boost::regex::extended);

	if (regex_search(s, what, add_target, boost::match_continuous)) {
	    cmd_do4(sock, "add_target", what[1], what[2], what[3]);
	    continue;
	}

	static const boost::regex 
	    remove_target(" *remove +target +([^ ]+) +([^ ]+) *$", 
			  boost::regex::extended);
	
	if (regex_search(s, what, remove_target, boost::match_continuous)) {
	    cmd_do3(sock, "remove_target", what[1], what[2]);
	    continue;
	}

	static const boost::regex 
	    add_parameter(" *add +parameter +([^ ]+) +(.*) *$", 
			  boost::regex::extended);

	if (regex_search(s, what, add_parameter, boost::match_continuous)) {
	    cmd_do3(sock, "add_parameter", what[1], what[2]);
	    continue;
	}

	static const boost::regex 
	    remove_parameter(" *remove +parameter +([^ ]+) *$", 
			  boost::regex::extended);
	
	if (regex_search(s, what, remove_parameter, boost::match_continuous)) {
	    cmd_do2(sock, "remove_parameter", what[1]);
	    continue;
	}

	static const boost::regex 
	    add_endpoint(" *add +endpoint +([^ ]+) +([^ ]+) +([^ ]+) +([^ ]+) *$", 
			  boost::regex::extended);

	if (regex_search(s, what, add_endpoint, boost::match_continuous)) {
	    cmd_do5(sock, "add_endpoint", what[1], what[2], what[3],
		    what[4]);
	    continue;
	}

	static const boost::regex 
	    remove_endpoint(" *remove +endpoint +([^ ]+) +([^ ]+) +([^ ]+) +([^ ]+) *$", 
			  boost::regex::extended);
	
	if (regex_search(s, what, remove_endpoint, boost::match_continuous)) {
	    cmd_do5(sock, "remove_endpoint", what[1], what[2], what[3],
		    what[4]);
	    continue;
	}

	std::cerr << "Command not understood." << std::endl;
	continue;

    }

}

int main(int argc, char** argv)
{
    try {
	client(argc, argv);
    } catch (std::exception& e) {
	std::cerr << "Exception: " << e.what() << std::endl;
    }
    exit(0);
}
